### DISCUSSION TITLE
Filter applied on $expand clause

### DISCUSSION THREAD

**Discussion by imoraca** [AutoMapper/AutoMapper.Extensions.OData#127] [Q&A] [answered]
Hi guys,

You're doing a wonderful job here, but I think I may have a little problem.

I'm trying to apply a filter on expand clause but it gets ignored. The query is something like this

http://localhost:<port>/opstenant?$expand=Buildings($filter=Prop1 eq Val1)

Any ideas.

**Comment by BlaiseD** [comment, marked as answer]
The [tests](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/0e90f4ccd80d212540b0443d92e7ecbb0829c4a8/AutoMapper.OData.EFCore.Tests/GetQueryTests.cs#L837) seem to work.  Maybe add a failing test reproducing your failed case?

**Reply by imoraca** [reply]
Hi there,

Yes, your example works and I figured out why mine didn't work. The only difference is that I used GetAsync method instead of GetQueryAsync.

The code that didn't work:
return Ok(await Context.BuildingSet.GetAsync(_mapper, options));

The code that works:
return Ok(await Context.BuildingSet.GetQueryAsync(_mapper, options));

I appreciate your help. This saved me a lot of time.

**Reply by BlaiseD** [reply]
Yes - that's correct.  `GetQuery` uses projection where we have an implementation for this feature.  `Get` uses includes where we don't.  PRs welcome.

**Reply by imoraca** [reply]
I also noticed that Get has better performance. It produces simpler SQL, although It gets all data for expanded properties ignoring filter.  So now I'm trying something like this:

            var coll = await query.GetAsync(_mapper, queryOptions);
            var newQuery = queryOptions.ApplyTo(coll.AsQueryable());
            return Ok(newQuery);

I'm aware that more data is loaded into memory this way and then on that loaded data I'm using queryOptions.ApplyTo method, but this way filter on expand clause is applied and I get good performance. 

Is this ok, or do you have some other suggestions?
